snuba/web/rpc/v1/resolvers/common/cross_item_queries.py

```diff
@@ -0,0 +1,151 @@
+from google.protobuf.json_format import MessageToDict
+from proto import Message  # type: ignore
+from sentry_protos.snuba.v1.endpoint_get_traces_pb2 import GetTracesRequest
+from sentry_protos.snuba.v1.request_common_pb2 import (
+    RequestMeta,
+    TraceItemFilterWithType,
+)
+
+from snuba import state
+from snuba.attribution.appid import AppID
+from snuba.attribution.attribution_info import AttributionInfo
+from snuba.datasets.entities.entity_key import EntityKey
+from snuba.datasets.entities.factory import get_entity
+from snuba.datasets.pluggable_dataset import PluggableDataset
+from snuba.query import SelectedExpression
+from snuba.query.data_source.simple import Entity
+from snuba.query.dsl import Functions as f
+from snuba.query.dsl import and_cond, column, or_cond
+from snuba.query.logical import Query
+from snuba.query.query_settings import HTTPQuerySettings
+from snuba.request import Request as SnubaRequest
+from snuba.utils.metrics.timer import Timer
+from snuba.web.query import run_query
+from snuba.web.rpc.common.common import (
+    base_conditions_and,
+    trace_item_filters_to_expression,
+    treeify_or_and_conditions,
+)
+from snuba.web.rpc.common.debug_info import setup_trace_query_settings
+from snuba.web.rpc.v1.resolvers.R_eap_items.common.common import (
+    attribute_key_to_expression,
+)
+
+# This is a hack to allow for the use of both TraceItemFilterWithType and GetTracesRequest.TraceFilter in the same function.
+JointTraceFilterType = (
+    RepeatedCompositeFieldContainer[TraceItemFilterWithType]
+    | RepeatedCompositeFieldContainer[GetTracesRequest.TraceFilter]
+)
+
+
+def get_trace_ids_for_cross_item_query(
+    original_request: Message,
+    request_meta: RequestMeta,
+    trace_filters: JointTraceFilterType,
+    timer: Timer,
+) -> list[str]:
+    """
+    This function is used to get the trace ids that match the given trace filters.
+    It does this by creating a query that looks like this:
+    SELECT trace_id FROM eap_items
+    WHERE (item_type = <item_type_0> AND <trace_item_filter_expression_0>) OR ... OR (item_type = <item_type_n> AND <trace_item_filter_expression_n>)
+    GROUP BY trace_id
+    HAVING countIf(item_type = <item_type_0> AND <trace_item_filter_expression_0>) > 0 AND ... AND countIf(item_type = <item_type_n> AND <trace_item_filter_expression_n>) > 0
+
+    This works by pruning out items that don't match any of the conditions in the where close. The HAVING
+    clause is used to get trace ids that contains items matching all of the conditions.
+    """
+    assert len(trace_filters) > 1, "At least two item types are required for a cross-event query"
+
+    # Hacky conversion due to protobuf ugliness
+    converted_trace_filters = [trace_filter for trace_filter in trace_filters]
+    if isinstance(trace_filters[0], GetTracesRequest.TraceFilter):
+        converted_trace_filters = [
+            TraceItemFilterWithType(item_type=trace_filter.item_type, filter=trace_filter.filter)
```

Comment:
Maybe we want to do the conversion earlier so this function sticks with handling one type only since, in the end, that conversion is endpoint dependent and will be removed.